hail/python/hail/linalg/blockmatrix.py

```diff
@@ -2169,14 +2035,78 @@ def _svd_gramian(self, compute_uv):
 block_matrix_type.set(BlockMatrix)
 
 
-def _shape(b):
-    if isinstance(b, int) or isinstance(b, float):
-        return 1, 1
-    if isinstance(b, np.ndarray):
-        b = _ndarray_as_2d(b)
+def _is_scalar(x):
+    return isinstance(x, float) or isinstance(x, int)
+
+
+def _shape(x):
+    if _is_scalar(x):
+        return []
+
+    return x.shape
+
+
+def _shape_after_broadcast(left_shape, right_shape):
+    """
+    Follows numpy's strategy of broadcasting through right-align shapes and
+    compare corresponding dimensions. See:
+    https://docs.scipy.org/doc/numpy-1.15.0/user/basics.broadcasting.html#general-broadcasting-rules
+    """
+    def _calc_new_dim(l_dim_length, r_dim_length):
+        if not (l_dim_length == r_dim_length or l_dim_length == 1 or r_dim_length == 1):
+            raise ValueError(f'Incompatible shapes for broadcasting: {left_shape}, {right_shape}')
+
+        return max(l_dim_length, r_dim_length)
+
+    left_ndim, right_ndim = len(left_shape), len(right_shape)
+    if left_ndim < right_ndim:
+        left_shape = [1 for _ in range(right_ndim - left_ndim)] + left_shape
+    elif right_ndim < left_ndim:
+        right_shape = [1 for _ in range(left_ndim - right_ndim)] + right_shape
+
+    return [_calc_new_dim(l, r) for l, r in zip(left_shape, right_shape)]
+
+
+def _to_bmir(x):
+    if _is_scalar(x):
+        return ValueToBlockMatrix(F64(x), hl.tfloat64, [], BlockMatrix.default_block_size(), [])
+    elif isinstance(x, np.ndarray):
+        return ValueToBlockMatrix(_ndarray_to_makearray(x), hl.tfloat64, list(x.shape),
+                                  BlockMatrix.default_block_size(), [False for _ in x.shape])
```

Comment:
I think for now this is OK. I think we should move towards not exposing block size to the user at all. The system should be free to choose a block_size amenable to the computation as a whole.

Reply:
Ok. I agree block size doesn't seem to make much sense for the user to be aware of. Especially if we can choose a better block size for them.

Reply:
The main issue is that if a user tries to do 
```BlockMatrix.from_numpy(ndarray, block_size=2) + ndarray``` it will complain about block size not being equal and the implicit conversion to a BlockMatrix with a default block size is not very clear (nor should they be concerned about that)

Reply:
Either way there's some form of intermediate awkwardness before `block_size` is hidden from the user. Does it even make sense to consider block size at this level? Is this something that can  be determined much later on, especially if we will ultimately not be able to know dimension-size statically?

Reply:
**Consensus:** will use block size from BlockMatrix operand from the binary operation.